### ISSUE TITLE
vg construction, Segmentation fault 

### ISSUE THREAD

**Issue by evanbiederstedt** [vgteam/vg#1647]
It's possible this is an older version of vg---I'll update with the new installation which I'm trying

```
$ vg construct -r hg38.fa -v myfile.vcf.gz  > output.vg

Stack trace (most recent call last):
#9    Object "", at 0xffffffffffffffff, in 
#8    Object "/nfs/sw/vg/vg-1.6.1/vg", at 0x6b63ed, in _start
#7    Object "/usr/lib64/libc-2.17.so", at 0x7f4995da0b14, in __libc_start_main
#6    Source "src/main.cpp", line 61, in main [0x5dd264]
#5    Source "src/subcommand/construct_main.cpp", line 287, in main_construct [0xa55e88]
#4    Source "src/constructor.cpp", line 1471, in construct_graph [0xac8fa4]
#3    Source "src/constructor.cpp", line 1306, in construct_graph [0xac813f]
#2    Source "src/constructor.cpp", line 437, in construct_chunk [0xac619a]
#1    Object "/nfs/sw/vg/vg-1.6.1/vg", at 0xd834b3, in vcflib::Variant::parsedAlternates(bool, bool, bool, float, float, float, float, float, std::string, std::string)
#0    Object "/nfs/sw/vg/vg-1.6.1/vg", at 0xd9407c, in CSmithWatermanGotoh::Align(unsigned int&, std::string&, std::string const&, std::string const&)
Segmentation fault (Address not mapped to object [0x590246000])
Segmentation fault
```

Is this an error seen previously which has been fixed circa March 2018?

**Comment by adamnovak** [comment]
I've never seen this before. It looks like an internal crash in vcflib, when trying to interpret the alts for one of your VCF records (and align them to each other and the reference in a sensible way).

Is your VCF file unusual? Maybe it is doing something with the alts that vcflib can't deal with.

**Comment by evanbiederstedt** [comment]
@adamnovak 

> Is your VCF file unusual? Maybe it is doing something with the alts that vcflib can't deal with.

The ALTs are quite long. I'm filtering the VCF based on line size to check whether this is why vcflib crashes. 

You're happy to have a look of course

**Comment by evanbiederstedt** [comment]
@edawson Thanks for the help. 

     -f, --flat-alts N     don't chop up alternate alleles from input vcf

I'm a bit confused how this flag should work...How do I use this with 

    $ vg construct -r hg38.fa -v myfile.vcf.gz  > output.vg

**Comment by edawson** [comment]
Sure thing - `-f` can go in your command like this:  
`$ vg construct -f -r hg38.fa -v myfile.vcf.gz  > output.vg `

This will assume that all your variants are perfectly canonical VCF and make them simple bubbles in the graph (ref on one side of the bubble and the alt on the other). It might, however, mess up things like left-align normalization, multiple alleles, etc if the variants in the VCF aren't pretty.

**Comment by evanbiederstedt** [comment]
> Sure thing - -f can go in your command like this:

Excellent, thanks. I was getting errors with the arguments in a different order, i.e. `-r -f`

**Comment by edawson** [comment]
Ah, yes in that case vg would think that your fasta reference was named "-f". I assume the error you got wasn't so informative - any chance you could paste it here so we could track it down?

**Comment by evanbiederstedt** [comment]
@edawson 

Here is the error I get running the command

    $ vg construct -f -r hg38.fa -v myfile.vcf.gz > output.vg

It happens within seconds

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  stream::write: message too large error writing protobuf
Stack trace (most recent call last):
#17   Object "", at 0xffffffffffffffff, in 
#16   Object "/nfs/sw/vg/vg-1.6.1/vg", at 0x6b63ed, in _start
#15   Object "/usr/lib64/libc-2.17.so", at 0x7fc61b5beb14, in __libc_start_main
#14   Source "src/main.cpp", line 61, in main [0x5dd264]
#13   Source "src/subcommand/construct_main.cpp", line 287, in main_construct [0xa55e88]
#12   Source "src/constructor.cpp", line 1471, in construct_graph [0xac8fa4]
#11   Source "src/constructor.cpp", line 1309, in construct_graph [0xac81cd]
#10 | Source "src/constructor.cpp", line 1196, in operator()
      Source "/nfs/sw/gcc/gcc-4.9.2/rtf/include/c++/4.9.2/functional", line 2439, in operator() [0xac7041]
       2436:     {
       2437:       if (_M_empty())
       2438: 	__throw_bad_function_call();
      >2439:       return _M_invoker(_M_functor, std::forward<_ArgTypes>(__args)...);
       2440:     }
       2441: 
       2442: #ifdef __GXX_RTTI
#9  | Source "/nfs/sw/gcc/gcc-4.9.2/rtf/include/c++/4.9.2/functional", line 2039, in operator()
    |  2037:       _M_invoke(const _Any_data& __functor, _ArgTypes... __args)
    |  2038:       {
    | >2039: 	(*_Base::_M_get_pointer(__functor))(
    |  2040: 	    std::forward<_ArgTypes>(__args)...);
    |  2041:       }
      Source "src/subcommand/construct_main.cpp", line 263, in _M_invoke [0xa5698f]
#8    Source "/nfs/sw/src/vg-0208/vg/src/stream.hpp", line 138, in write<vg::Graph> [0xa3e9c9]
        135:         handle(!coded_out.HadError());
        136:     }
        137: 
      > 138:     return !count || written == count;
        139: }
        140: 
        141: template <typename T>
#7    Source "/nfs/sw/gcc/gcc-4.9.2/src/gcc-4.9.2/libgcc/unwind.inc", line 230, in _Unwind_Resume [0x7fc61b96f4f6]
        227:   /* Choose between continuing to process _Unwind_RaiseException
        228:      or _Unwind_ForcedUnwind.  */
        229:   if (exc->private_1 == 0)
      > 230:     code = _Unwind_RaiseException_Phase2 (exc, &cur_context);
        231:   else
        232:     code = _Unwind_ForcedUnwind_Phase2 (exc, &cur_context);
#6    Source "/nfs/sw/gcc/gcc-4.9.2/src/gcc-4.9.2/libgcc/unwind.inc", line 62, in _Unwind_RaiseException_Phase2 [0x7fc61b96efd2]
         59:       /* Unwind successful.  Run the personality routine, if any.  */
         60:       if (fs.personality)
         61: 	{
      >  62: 	  code = (*fs.personality) (1, _UA_CLEANUP_PHASE | match_handler,
         63: 				    exc->exception_class, exc, context);
         64: 	  if (code == _URC_INSTALL_CONTEXT)
         65: 	    break;
#5    Source "/nfs/sw/gcc/gcc-4.9.2/src/gcc-4.9.2/libstdc++-v3/libsupc++/eh_personality.cc", line 670, in __gxx_personality_v0 [0x7fc61c0ea99a]
        667:   else
        668:     {
        669:       if (found_type == found_terminate)
      > 670: 	__cxa_call_terminate(ue_header);
        671: 
        672:       // Cache the TType base value for __cxa_call_unexpected, as we won't
        673:       // have an _Unwind_Context then.
#4    Source "/nfs/sw/gcc/gcc-4.9.2/src/gcc-4.9.2/libstdc++-v3/libsupc++/eh_call.cc", line 54, in __cxa_call_terminate [0x7fc61c0ea1d8]
         51: 	  __cxa_exception* xh;
         52: 
         53: 	  xh = __get_exception_header_from_ue(ue_header);
      >  54: 	  __terminate(xh->terminateHandler);
         55: 	}
         56:     }
         57:   /* Call the global routine if we don't have anything better.  */
#3    Source "/nfs/sw/gcc/gcc-4.9.2/src/gcc-4.9.2/libstdc++-v3/libsupc++/eh_terminate.cc", line 47, in __terminate [0x7fc61c0eb145]
         44: {
         45:   __try 
         46:     {
      >  47:       handler ();
         48:       std::abort ();
         49:     } 
         50:   __catch(...) 
#2    Source "/nfs/sw/gcc/gcc-4.9.2/src/gcc-4.9.2/libstdc++-v3/libsupc++/vterminate.cc", line 95, in __verbose_terminate_handler [0x7fc61c0ed0dc]
         92:     else
         93:       fputs("terminate called without an active exception\n", stderr);
         94:     
      >  95:     abort();
         96:   }
         97: 
         98: _GLIBCXX_END_NAMESPACE_VERSION
#1    Object "/usr/lib64/libc-2.17.so", at 0x7fc61b5d3ce7, in abort
#0    Object "/usr/lib64/libc-2.17.so", at 0x7fc61b5d25f7, in raise
Aborted (Signal sent by tkill() 138351 50168)
Aborted

```

**Comment by evanbiederstedt** [comment]
@edawson 

As for this command:

    $ vg construct -f -r hg38.fa -v myfile.vcf.gz > output.vg

I get this

`could not open -f`

**Comment by evanbiederstedt** [comment]
@adamnovak @edawson I'm happy to share the VCFs which cause problems---you could download from our servers.

**Comment by adamnovak** [comment]
Hm, this looks tricky. Protobuf has an upper bound on "message" size, which ends up being an upper bound on the size of any given "chunk" of our graphs. And vg generates constructed graphs chunk by chunk. And we just glom a certain number of variants into each chunk, without regard to how big the chunk is going to get as a result.

The workaround here is probably to pass `-z 100` or something even smaller, to limit the number of variants per chunk. But if a variant is sufficiently long, or if variants overlap, this may not work (since all of each variant, and all mutually overlapping variants, have to be generated in the same chunk).

The real fix for this is to add code to re-chunk the chunks that vg construct is generating, to ensure they don't get too big.

**Comment by apregier** [comment]
I'm having the same issue making graphs with large deletions (even with `-z 1`).  I'm happy to attempt a fix but not sure where to start - seems like I should be able to call PathChunker somehow from the callback in construct_main, but we don't have the xg index yet when constructing the graph, right?

**Comment by adamnovak** [comment]
I don't think the PathChunker or any of the fancy machinery in `vg chunk`
is needed here. (Those "chunks" are for variant calling, and partially
overlapping. The "chunks" we are worried about here are really just
incedental to the serialization.) The problem is that some of the Graph
objects being produced here have too many Node and Edge objects, and/or too
many Mappings in their Path objects, and are thus too big for Protobuf to
serialize.

The fix is to look at the Graph objects and work out which are too big, and
then to split the Node, Edge, and Path/Mapping objects up over multiple
Graph objects. I don't *think* there's any real worry about
Nodes/Edges/Mappings having to come in a particular order (i.e. you can
have an Edge before the Nodes it connects), but the Mappings within a Path
have to stay in the same relative order. In the limit, every Node, Edge,
and Mapping could be wrapped in its own Graph object, but that would waste
some space and serialization/deserialization time.

To put it another way, since we aren't using a vg::VG object, we need to
replicate its ability
https://github.com/vgteam/vg/blob/master/src/vg.cpp#L350 to break up the
graph into reasonably-sized chunks.

**Comment by adamnovak** [comment]
If we put the whole graph in a vg::VG object, we would have to build it all
in memory at once before writing it out. Doing it the way it is done right
now allows us to construct the graph in a streaming fashion, using a
bounded amount of memory.
